lto0.test_wasm_worker_futex_wait failure

```
C:\emsdk\emscripten\main>test\runner lto0.test_wasm_worker_futex_wait
Running 1 tests
[1/1] test_wasm_worker_futex_wait (test_core.lto0.test_wasm_worker_futex_wait) ... FAIL

Stdout:
-- begin program output --
Worker sleeping on futex...
Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
worker sent an error! Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
node:internal/event_target:1101
  process.nextTick(() => { throw err; });
                           ^

RuntimeError: Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
    at abort (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:739:11)
    at ___assert_fail (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1320:7)
    at wasm_worker_futex_wait.wasm._emscripten_check_timers (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[78]:0x3d3d)
    at wasm_worker_futex_wait.wasm._emscripten_yield (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[75]:0x3c5d)
    at wasm_worker_futex_wait.wasm._do_futex_wait (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[72]:0x3a41)
    at wasm_worker_futex_wait.wasm.emscripten_futex_wait (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[69]:0x390c)
    at wasm_worker_futex_wait.wasm.worker_main (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[68]:0x3799)
    at C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1063:69
    at callUserCallback (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1037:16)
    at _wasmWorkerRunPostMessage (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1063:19)
    at MessagePort.f (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:582:20)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
Thrown at:
    at node:internal/event_target:1101:28
    at processTicksAndRejections (node:internal/process/task_queues:85:11)


Node.js v22.16.0
-- end program output --


======================================================================
FAIL: test_wasm_worker_futex_wait (test_core.lto0.test_wasm_worker_futex_wait)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\emsdk\emscripten\main\test\test_core.py", line 130, in decorated
    return func(self, *args, **kwargs)
  File "C:\emsdk\emscripten\main\test\test_core.py", line 382, in decorated
    return func(self, *args, **kwargs)
  File "C:\emsdk\emscripten\main\test\test_core.py", line 171, in decorated
    return func(self, *args, **kwargs)
  File "C:\emsdk\emscripten\main\test\test_core.py", line 9728, in test_wasm_worker_futex_wait
    self.do_runf('wasm_worker/wasm_worker_futex_wait.c', cflags=['-sWASM_WORKERS'])
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\emsdk\emscripten\main\test\common.py", line 1387, in do_runf
    return self._build_and_run(filename, expected_output, **kwargs)
           ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\emsdk\emscripten\main\test\common.py", line 1435, in _build_and_run
    js_output = self.run_js(js_file, engine, args,
                            input=input,
                            assert_returncode=assert_returncode,
                            interleaved_output=interleaved_output)
  File "C:\emsdk\emscripten\main\test\common.py", line 1019, in run_js
    self.fail('JS subprocess failed (%s): %s (expected=%s).  Output:\n%s' % (error.cmd, error.returncode, assert_returncode, ret))
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: JS subprocess failed (C:/emsdk/node/22.16.0_64bit/bin/node.exe --stack-trace-limit=50 --trace-uncaught C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js): 1 (expected=0).  Output:
Worker sleeping on futex...
Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
worker sent an error! Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
node:internal/event_target:1101
  process.nextTick(() => { throw err; });
                           ^

RuntimeError: Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
    at abort (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:739:11)
    at ___assert_fail (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1320:7)
    at wasm_worker_futex_wait.wasm._emscripten_check_timers (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[78]:0x3d3d)
    at wasm_worker_futex_wait.wasm._emscripten_yield (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[75]:0x3c5d)
    at wasm_worker_futex_wait.wasm._do_futex_wait (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[72]:0x3a41)
    at wasm_worker_futex_wait.wasm.emscripten_futex_wait (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[69]:0x390c)
    at wasm_worker_futex_wait.wasm.worker_main (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[68]:0x3799)
    at C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1063:69
    at callUserCallback (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1037:16)
    at _wasmWorkerRunPostMessage (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1063:19)
    at MessagePort.f (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:582:20)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
Thrown at:
    at node:internal/event_target:1101:28
    at processTicksAndRejections (node:internal/process/task_queues:85:11)


Node.js v22.16.0


Stdout:
-- begin program output --
Worker sleeping on futex...
Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
worker sent an error! Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
node:internal/event_target:1101
  process.nextTick(() => { throw err; });
                           ^

RuntimeError: Aborted(Assertion failed: emscripten_is_main_runtime_thread(), at: /emsdk/emscripten/system/lib/libc/musl/src/signal/setitimer.c,71,_emscripten_check_timers)
    at abort (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:739:11)
    at ___assert_fail (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1320:7)
    at wasm_worker_futex_wait.wasm._emscripten_check_timers (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[78]:0x3d3d)
    at wasm_worker_futex_wait.wasm._emscripten_yield (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[75]:0x3c5d)
    at wasm_worker_futex_wait.wasm._do_futex_wait (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[72]:0x3a41)
    at wasm_worker_futex_wait.wasm.emscripten_futex_wait (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[69]:0x390c)
    at wasm_worker_futex_wait.wasm.worker_main (wasm://wasm/wasm_worker_futex_wait.wasm-00152626:wasm-function[68]:0x3799)
    at C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1063:69
    at callUserCallback (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1037:16)
    at _wasmWorkerRunPostMessage (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:1063:19)
    at MessagePort.f (C:\emsdk\emscripten\main\out\test\wasm_worker_futex_wait.js:582:20)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
Thrown at:
    at node:internal/event_target:1101:28
    at processTicksAndRejections (node:internal/process/task_queues:85:11)


Node.js v22.16.0
-- end program output --

----------------------------------------------------------------------
Ran 1 test in 1.874s

FAILED (failures=1)
```

Bisected down to

```
ef94d08193ff9dcfd985e4d2d1e674be8bc29cb1 is the first bad commit
commit ef94d08193ff9dcfd985e4d2d1e674be8bc29cb1
Author: Sam Clegg <sbc@chromium.org>
Date:   Mon Apr 13 22:36:02 2026 -0700

    [WASM_WORKERS] Fix UB in pthread_mutex with debug+hybrid builds (#26673)

    This fix only applies to the hybrid Wasm Workers + pthread build mode.

    We can't currently rely on calling `pthread_self()` from a Wasm Worker,
    even the hybrid mode. We are working on fixing that in #26631. However,
    we do have a `CURRENT_THREAD_ID` macro that works on both Wasm Workers
    and pthreads.

    As it happens, pthread locks under musl do not normally depend on
    `pthread_self` except for "recursive" which need to access the
    `pthread_self()->tid` field to track the current owner. This means that
    I broke the debug + hybrid mode locks in #24607 because I added the
    recursion check to all locks in debug builds.

    While I have added a repro case for this its not directly testing the
    issue, it just heppens to crash due to the current undefined behavior
    when doing `pthread_self()->tid` from a Wasm Worker.

    Fixes: #26619

 .../musl/src/thread/pthread_mutex_consistent.c     |  2 +-
 .../libc/musl/src/thread/pthread_mutex_timedlock.c |  4 +--
 .../libc/musl/src/thread/pthread_mutex_trylock.c   |  2 +-
 test/codesize/test_codesize_minimal_pthreads.json  |  4 +--
 .../test_codesize_minimal_pthreads_memgrowth.json  |  4 +--
 test/test_stress.py                                |  3 ++
 test/test_stress_pthread_malloc.c                  | 36 ++++++++++++++++++++++
 7 files changed, 47 insertions(+), 8 deletions(-)
 create mode 100644 test/test_stress_pthread_malloc.c
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -706,6 +706,7 @@ jobs:
             lto0.test_exceptions_allowed_uncaught
             lto0.test_longjmp_standalone_standalone
             lto0.test_embind_i64_val
+            lto0.test_wasm_worker_futex_wait
             thinlto0.test_pthread_dlsym
             core3
             core2g.test_externref

--- a/system/lib/libc/emscripten_yield_stub.c
+++ b/system/lib/libc/emscripten_yield_stub.c
@@ -6,6 +6,7 @@
  */
 
 #include <features.h>
+#include <emscripten/threading.h>
 
 #include "threading_internal.h"
 
@@ -16,5 +17,8 @@ static bool dummy(double now) {
 weak_alias(dummy, _emscripten_check_timers);
 
 bool _emscripten_yield(double now) {
-  return _emscripten_check_timers(now);
+  if (emscripten_is_main_runtime_thread()) {
+    return _emscripten_check_timers(now);
+  }
+  return false;
 }

--- a/test/codesize/test_minimal_runtime_code_size_hello_wasm_worker.json
+++ b/test/codesize/test_minimal_runtime_code_size_hello_wasm_worker.json
@@ -3,8 +3,8 @@
   "a.html.gz": 355,
   "a.js": 952,
   "a.js.gz": 601,
-  "a.wasm": 2652,
-  "a.wasm.gz": 1481,
-  "total": 4119,
-  "total_gz": 2437
+  "a.wasm": 2699,
+  "a.wasm.gz": 1510,
+  "total": 4166,
+  "total_gz": 2466
 }

